src/test/deploy/functions/cache/applyHash.spec.ts

```diff
@@ -0,0 +1,118 @@
+            functionsSourceV1: "backend1_sourceV1",
+            functionsSourceV2: "backend1_sourceV2",
+          },
+          backend2: {
+            functionsSourceV1: "backend2_sourceV1",
+            functionsSourceV2: "backend2_sourceV2",
+          },
+        },
+      };
+      const endpoint1 = {
+        ...EMPTY_ENDPOINT,
+        id: "endpoint1",
+        platform: "gcfv1",
+        codebase: "backend1",
+        secretEnvironmentVariables: [
+          {
+            key: "key",
+            secret: "secret1",
+            projectId: "projectId",
+            version: "1",
+          },
+        ],
+      } as backend.Endpoint;
+      const endpoint2 = {
+        ...EMPTY_ENDPOINT,
+        id: "endpoint2",
+        platform: "gcfv2",
+        codebase: "backend2",
+        secretEnvironmentVariables: [
+          {
+            key: "key",
+            secret: "secret2",
+            projectId: "projectId",
+            version: "2",
+          },
+        ],
+      } as backend.Endpoint;
+
+      const backend1 = backend.of(endpoint1);
+      const backend2 = backend.of(endpoint2);
+
+      backend1.environmentVariables.test = "shared_environment_variables";
+      backend2.environmentVariables = backend1.environmentVariables;
+
+      const backends = { backend1, backend2 };
+
+      const getSourceHash = sinon.stub(hash, "getSourceHash");
+      getSourceHash.callsFake((path: string) => Promise.resolve("source=" + path));
+
+      const getEnvironmentVariablesHash = sinon.stub(hash, "getEnvironmentVariablesHash");
+      getEnvironmentVariablesHash.callsFake(
+        (backend: backend.Backend) => "env=" + backend.environmentVariables.test
+      );
+      const getSecretsHash = sinon.stub(hash, "getSecretsHash");
+      getSecretsHash.callsFake(
+        (endpoint: backend.Endpoint) => "secret=" + endpoint.secretEnvironmentVariables?.[0].secret
+      );
+
+      const getEndpointHash = sinon.stub(hash, "getEndpointHash");
+      getEndpointHash.callsFake((source?: string, env?: string, secrets?: string) =>
+        [source, env, secrets].join("&")
+      );
+
+      // Execute
+      await applyBackendHashToBackends(backends, context);
+
+      // Expect
+      expect(getEndpointHash).to.have.been.calledWith(
+        "source=backend1_sourceV1",
+        "env=shared_environment_variables",
+        "secret=secret1"
+      );
+      expect(endpoint1.hash).to.equal(
+        "source=backend1_sourceV1&env=shared_environment_variables&secret=secret1"
+      );
+      expect(getEndpointHash).to.have.been.calledWith(
+        "source=backend2_sourceV2",
+        "env=shared_environment_variables",
+        "secret=secret2"
+      );
+      expect(endpoint2.hash).to.equal(
+        "source=backend2_sourceV2&env=shared_environment_variables&secret=secret2"
+      );
+      // Note: env variables only derived from the first backend
+      // (since all backends in wantBackends shared the same environment variables)
```

Comment:
Commented above - I don't think this is true. Each codebase can have its own set of environment variable.

Reply:
Thanks for calling out, removed comment and updated test!